stock_picking_type_routing_operation/models/stock_move.py

```diff
@@ -0,0 +1,315 @@
+# Copyright 2019 Camptocamp (https://www.camptocamp.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from itertools import chain
+from odoo import models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _action_assign(self):
+        super()._action_assign()
+        if not self.env.context.get("exclude_apply_routing_operation"):
+            self._apply_src_move_routing_operation()
+            self._apply_dest_move_routing_operation()
+
+    def _apply_src_move_routing_operation(self):
+        src_moves = self._split_per_src_routing_operation()
+        src_moves._apply_move_location_src_routing_operation()
+
+    def _apply_dest_move_routing_operation(self):
+        dest_moves = self._split_per_dest_routing_operation()
+        dest_moves._apply_move_location_dest_routing_operation()
+
+    def _split_per_src_routing_operation(self):
+        """Split moves per source routing operations
+
+        When a move has move lines with different routing operations or lines
+        with routing operations and lines without, on the source location, this
+        method split the move in as many source routing operations they have.
+
+        The reason: the destination location of the moves with a routing
+        operation will change and their "move_dest_ids" will be modified to
+        target a new move for the routing operation.
+        """
+        move_to_assign_ids = set()
+        new_move_per_location = {}
+        for move in self:
+            if move.state not in ("assigned", "partially_available"):
+                continue
+
+            # Group move lines per source location, some may need an additional
+            # operations while others not. Store the number of products to
+            # take from each location, so we'll be able to split the move
+            # if needed.
+            move_lines = {}
+            for move_line in move.move_line_ids:
+                location = move_line.location_id
+                move_lines[location] = sum(move_line.mapped("product_uom_qty"))
```

Comment:
@guewen 
Here we consider `product_uom_qty`, but later in the split, Odoo expects the quantity in default product uom.
I think we should consider the `product_qty` instead.

I made a quick  test, and It seems it ends on working, but not the way we expect.
I have created a product A with 6 qty in WH/Stock/highbay and 18 in WH/Stock/shelf 1
I left default uom as unit.
Then, I have created a delivery order with qty 2 / product_uom dozen

After reservation, I have
The Highbay=>Handover with 2 moves with unit as uom (with qty 0.5 and 5.5)
Then I also have a delivery order with one move, with dozen as uom and qty 1.5.
Well, at last, it is all kind of good, but it is weird that the uom of stock move changes during reservation, and it is weird that we have 2 moves with qty 0.5 and 5.5 in the same picking.

I guess if we work with qty in product uom, it will resolve all of this.
I did not test, but I guess something similar happens in `_split_per_dest_routing_operation`